README.md
SDRdaemon
=========

**SDRdaemon** package can be used to:
  - send I/Q samples read from a SDR device over the network via UDP (Rx mode)
  - rceive I/Q samples from the network via UDP and send them to a SDR device (Tx mode) 

<h1>Introduction</h1>

**sdrdaemonrx** is a basic software-defined radio receiver that just sends the I/Q samples over the network via UDP. It was developed on the base of NGSoftFM (also found in this Github repo: https://github.com/f4exb/ngsoftfm) and shares a lot of the code for the interface with the SDR hardware devices.

It conveys meta data in the data flow so that the receiving application is informed about parameters essential to render correctly the data coming next such as the sample rate, the number of bytes used for the samples, the number of effective sample bits, the center frequency... (See the "Data format" chapter for detals).

While running the program accepts configuration commands on a TCP port using nanomsg messages with a content in the same format as the configuration string given on the command line (See the "Running" chapter for details). This provides a dynamic control of the device or features of the application such as the decimation. A Python script is provided to send such messages.

In order to recover possible lost blocks it uses Cauchy MDS Block Erasure codec to encode data with redundancy, It can add a user defined number of redundant block so that if the nominal number of blocks is received (128 blocks) it can recover the lost blocks in any position.

Note that if you set the number of redundant blocks to 0 then no FEC is used.

**sdrdaemontx** does the same thing as sdrdaemonrx but the other way round. It takes blocks read from UDP possibly with redundant blocks and sends the I/Q data to a SDR transmitter.

Hardware supported:

Receivers:

  - **RTL-SDR** based (RTL2832-based) hardware is supported and uses the _librtlsdr_ library to interface with the RTL-SDR hardware.
  - **HackRF** One and variants are supported with _libhackrf_ library.
  - **Airspy** is supported with _libairspy_ library.
  - **BladeRF** is supported with _libbladerf_ library.

Transmitters:

  - **HackRF** One and variants are supported with _libhackrf_ library.

SDRdaemon programs can be used conveniently along with SDRangel (found in this Github repo: https://github.com/f4exb/sdrangel) as the client application. So in this remote type of configuration you will need both an angel and a daemon :-)

GNUradio is also supported with a specific `sdrdaemonsource`source block for Rx devices provided in the `gr-sdrdaemon` OOT module. The `sdrdaemonsink` sink block for Tx devices does not exist at the moment.

SDRdaemon package requires:

 - Linux
 - C++11
 - Boost includes only (library is not linked with SDRdaemon)
 - nanomsg
 - CM256cc library found in this Github repo: https://github.com/f4exb/cm256cc
 - RTL-SDR library (http://sdr.osmocom.org/trac/wiki/rtl-sdr) for RTL-SDR support
 - HackRF library (https://github.com/mossmann/hackrf/tree/master/host/libhackrf) for HackRF support
 - Airspy library (https://github.com/airspy/host/tree/master/libairspy) for Airspy support
 - BladeRF library (https://github.com/Nuand/bladeRF/tree/master/host) for BladeRF support
 - supported hardware
 - A computer or embedded device such as the Raspberry Pi 2 or 3 to which you connect the hardware.
 
&#9758; On Raspberry Pi 3 it is recommended to run in a native aarch64 system for better performance. This is particularly true on the Tx side with `sdrdaemontx`. For the moment only OpenSUSE does that: [link](https://en.opensuse.org/HCL:Raspberry_Pi3).

For the latest version, see https://github.com/f4exb/SDRdaemon

Branches:

  - _master_ is the "production" branch with the most stable release
  - _dev_ is the development branch that contains current developments that will be eventually released in the master branch
  - _fix_ contains fixes that cannot wait for the dev branch to go to production


<h1>Prerequisites</h1>

<h2>Base requirements</h2>

<h3>Ubuntu/Debian</h3>

  - `sudo apt-get install cmake pkg-config libusb-1.0-0-dev libasound2-dev libboost-all-dev libnanomsg-dev`

<h3>OpenSUSE</h3>

in particular if you run the aarch64 version with the RPi3. Performance in an aarch64 (ARM 64 bits amrv8) is considerably better than in an armv7 system. This is particularly true with `sdradaemontx`. Use gcc version 6 to get automatic vectorization.

  - `sudo zypper install cmake gcc6 gcc6-c++ libusb-1_0-devel boost-devel fftw3-devel`

Do once in the shell where you compile:

  - `export CC=/usr/bin/gcc-6`
  - `export CXX=/usr/bin/g++-6`

You will need to compile and install libnanomsg because it is not available as a package:

  - `git clone https://github.com/ZewoGraveyard/libnanomsg.git`
  - `cd libnanomsg`
  - `mkdir build; cd build`
  - `cmake -DCMAKE_INSTALL_PREFIX=/opt/install/libnanomsg ..`
  - `make -j4 install`

<h2>Forward Erasure Correction (FEC) support</h2>

You have to install [CM256cc](https://github.com/f4exb/cm256cc). You will then have to specify the include and library paths on the cmake command line. Say if you install cm256cc in `/opt/install/cm256cc` you will have to add `-DCM256CC_INCLUDE_DIR=/opt/install/cm256cc/include/cm256cc -DCM256CC_LIBRARIES=/opt/install/cm256cc/lib/libcm256cc.so` to the cmake commands.

`sdrdaemonrx` binary recognizes the configuration commmand `fecblk` to specify the number of FEC blocks. When `fecblk=0` is specified in the commands and hence no FEC blocks are added. 

Frames and blocks are numbered and even if no FEC blocks are added this can help in reconstructing frames with appropriate timings.

<h2>Airspy support</h2>

Airspy support must be installed for SDRdaemon to work with an Airspy device.

If you install from source (https://github.com/airspy/host/tree/master/libairspy) in your own installation path you have to specify the include path and library path. For example if you installed it in `/opt/install/libairspy` you have to add `-DLIBAIRSPY_LIBRARIES=/opt/install/libairspy/lib/libairspy.so -DLIBAIRSPY_INCLUDE_DIR=/opt/install/libairspy/include` to the cmake options.

To install the library from a Debian/Ubuntu installation just do:

  - `sudo apt-get install libairspy-dev`

<h2>BladeRF support</h2>

BladeRF support must be installed for SDRdaemon to work with a BladeRF device.

If you install from source (https://github.com/Nuand/bladeRF) in your own installation path you have to specify the include path and library path. For example if you installed it in `/opt/install/libbladerf` you have to add `-DLIBBLADERF_LIBRARIES=/opt/install/libbladeRF/lib/libbladeRF.so -DLIBBLADERF_INCLUDE_DIR=/opt/install/libbladeRF/include` to the cmake options.

To install the library from a Debian/Ubuntu installation just do:

  - `sudo apt-get install libbladerf-dev`

Note: for the BladeRF to work effectively on FM broadcast frequencies you have to fit it with the XB200 extension board.

<h2>HackRF support</h2>

HackRF support must be installed for SDRdaemon to work with a HackRF device.

If you install from source (https://github.com/mossmann/hackrf/tree/master/host/libhackrf) in your own installation path you have to specify the include path and library path. For example if you installed it in `/opt/install/libhackrf` you have to add `-DLIBHACKRF_LIBRARIES=/opt/install/libhackrf/lib/libhackrf.so -DLIBHACKRF_INCLUDE_DIR=/opt/install/libhackrf/include` to the cmake options.

To install the library from a Debian/Ubuntu installation just do:

  - `sudo apt-get install libhackrf-dev`

<h2>RTL-SDR support</h2>

The Osmocom RTL-SDR library must be installed before you can use SDRdaemon with a RTL-SDR device.
See http://sdr.osmocom.org/trac/wiki/rtl-sdr for more information.
SDRdaemon has been tested successfully with RTL-SDR 0.5.3. Normally your distribution should provide the appropriate librtlsdr package.
If you go with your own installation of librtlsdr you have to specify the include path and library path. For example if you installed it in `-DLIBRTLSDR_LIBRARIES=/opt/install/librtlsdr/lib/librtlsdr.so -DLIBRTLSDR_INCLUDE_DIR=/opt/install/librtlsdr/include` to the cmake options

To install the library from a Debian/Ubuntu installation just do:

  - `sudo apt-get install librtlsdr-dev`

<h2>nanomsg custom installation</h2>

If you build nanomsg from source obtained either by git clone or a released source package and install it in your own path (ex: `/opt/install/libnanomsg`) you will need to specify the include and library paths like this: `-DLIBNANOMSG_LIBRARIES=/opt/install/libnanomsg/lib/libnanomsg.so -DLIBNANOMSG_INCLUDE_DIR=/opt/install/libnanomsg/include`

<h1>Installing</h1>

To install SDRdaemon, download and unpack the source code and go to the
top level directory. Then do like this:

 - `mkdir build`
 - `cd build`
 - `cmake ..`

Compile and install

 - `make -j8` (for machines with 8 CPUs)
 - `make install`


<h1>Running</h1>

<h2>Rx examples</h2>

Typical commands:

  - RTL-SDR: `./sdrdaemonrx -t rtlsdr -I 192.168.1.3 -D 9090 -C 9091 -c txdelay=300,fecblk=8,freq=433970000,srate=1000000,ppmp=58,gain=40.2,decim=5,fcpos=2`
    - Use RTL-SDR device #0
    - Destination address for the data is: `192.168.1.3`
    - Using UDP port `9090` for the data (it is the default anyway)
    - Using TCP port `9091` to listen to configuration commands (it is the default anyway)
    - Startup configuration:
      - Wait between UDP blocks: 300 microseconds
      - FEC: add 8 FEC blocks to the 128 blocks data frame resulting in a total of 136 blocks per frame.
      - Center frequency: _433.97 MHz_
      - Device sample rate: _1 MHz_
      - Local oscillator correction: _58 ppm_
      - RF gain: _40.2 dB_
      - Decimation: 2^_5_ = 32; thus stream sample rate is 31.25 kHz
      - Position of center frequency: _2_ is centered (decimation around the center)
  - Airspy: `./sdrdaemonrx -t airspy -I 192.168.1.3 -D 9090 -c txdelay=300,fecblk=8,freq=433970000,srate=10000000,ppmn=1.7,lgain=13,mgain=9,vgain=6,decim=5,fcpos=0`
    - Use Airspy device #0
    - Destination address for the data is: `192.168.1.3`
    - Using UDP port `9090` for the data (it is the default anyway)
    - Using TCP port `9091` to listen to configuration commands (it is the default anyway)
    - Startup configuration:
      - Wait between UDP blocks: 300 microseconds
      - FEC: add 8 FEC blocks to the 128 blocks data frame resulting in a total of 136 blocks per frame.
      - Center frequency: _433.97 MHz_
      - Device sample rate: _10 MHz_
      - LO correction: _-1.7 ppm_
      - LNA gain: _13 dB_

debian/changelog
sdrdaemon (3.1.2-1) unstable; urgency=medium

  * RTLSDR and Airspy: added possibility to turn off the flags

 -- Edouard Griffiths, F4EXB <f4exb06@gmail.com>  Sat, 08 Jul 2017 09:14:18 +0200

sdrdaemon (3.1.1-1) unstable; urgency=medium

  * Fixed HacrRF Rx by using signed char for the HackRF samples

 -- Edouard Griffiths, F4EXB <f4exb06@gmail.com>  Fri, 30 Jun 2017 08:14:18 +0200

sdrdaemon (3.1.0-1) unstable; urgency=medium

  * Removed sdrdaemon from the build and liblz4 dependency
  * Removed gr-sdrdaemonfec  GNUradio OOT module
  * Created a gr-sdrdaemon GNUradio OOT module with sdrdaemonsource block for Rx
  * RTLSDR: use asynchronous reads. Fixes I/Q stream issues in particular with RPi hosts

 -- Edouard Griffiths, F4EXB <f4exb06@gmail.com>  Sun, 25 Jun 2017 23:14:18 +0200

sdrdaemon (3.0.1-1) unstable; urgency=medium

  * Fixed UDP transmission irregularities in Rx mode by adding buffering
  * Fixed bandwidth selection with the HackRF
  * Added file and documentation for service implementation in openSUSE

 -- Edouard Griffiths, F4EXB <f4exb06@gmail.com>  Sun, 18 Jun 2017 09:14:18 +0200

sdrdaemon (3.0.0-1) unstable; urgency=medium

  * Transmission support with HackRF
  * Also file sink for debugging
  * Renamed sdrdaemonfec to sdrdaemonrx

 -- Edouard Griffiths, F4EXB <f4exb06@gmail.com>  Sun, 11 Jun 2017 05:14:18 +0100

sdrdaemon (2.1.3-1) unstable; urgency=medium

  * Fixed errors in cmake when no cm256cc package is available
  * Fixed cmake file to have a library suffix
  * Fixed wrong usage of assert: not working for release builds
    
 -- Edouard Griffiths, F4EXB <f4exb06@gmail.com>  Wed, 26 Apr 2017 23:14:18 +0100

sdrdaemon (2.1.2-1) unstable; urgency=medium

  * Initial release as a Debian package

 -- Edouard Griffiths, F4EXB <f4exb06@gmail.com>  Mon, 16 Jan 2017 23:15:15 +0200


sdrdaemonrx.cpp
///////////////////////////////////////////////////////////////////////////////////
// SDRdaemon - send I/Q samples read from a SDR device over the network via UDP. //
//                                                                               //
// This version implements Forward Erasure Correction (FEC) to be able to        //
// recover lost blocks at the receiving end. It uses Cauchy Reed-Solomon coding  //
// for the redundant blocks and is based on the modified cm256 library (see:     //
// https://github.com/f4exb/cm256) with g++ cmake and Neon for armv7l devices    //
// support.                                                                      //
//                                                                               //
// Copyright (C) 2016 Edouard Griffiths, F4EXB                                   //
//                                                                               //
// This program is free software; you can redistribute it and/or modify          //
// it under the terms of the GNU General Public License as published by          //
// the Free Software Foundation as version 3 of the License, or                  //
//                                                                               //
// This program is distributed in the hope that it will be useful,               //
// but WITHOUT ANY WARRANTY; without even the implied warranty of                //
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the                  //
// GNU General Public License V3 for more details.                               //
//                                                                               //
// You should have received a copy of the GNU General Public License             //
// along with this program. If not, see <http://www.gnu.org/licenses/>.          //
///////////////////////////////////////////////////////////////////////////////////

#include <cstdlib>
#include <cstdio>
#include <climits>
#include <cmath>
#include <csignal>
#include <cstring>
#include <algorithm>
#include <atomic>
#include <memory>
#include <thread>
#include <unistd.h>
#include <getopt.h>
#include <sys/time.h>

#include "util.h"
#include "DataBuffer.h"
#include "Downsampler.h"
#include "UDPSinkFEC.h"

#ifdef HAS_RTLSDR
    #include "RtlSdrSource.h"
#endif
#ifdef HAS_HACKRF
    #include "HackRFSource.h"
#endif
#ifdef HAS_AIRSPY
    #include "AirspySource.h"
#endif
#ifdef HAS_BLADERF
    #include "BladeRFSource.h"
#endif
#include "TestSource.h"
#include "SDRDaemon.h"

//#include <type_traits>

#define UDPSIZE 512

/** Flag is set on SIGINT / SIGTERM. */
static std::atomic_bool stop_flag(false);


/** Simple linear gain adjustment. */
void adjust_gain(SampleVector& samples, double gain)
{
    for (unsigned int i = 0, n = samples.size(); i < n; i++) {
        samples[i] *= gain;
    }
}

/**
 * Get data from output buffer and write to output stream.
 *
 * This code runs in a separate thread.
 */
void write_output_data(UDPSink *output,
        DataBuffer<IQSample> *buf,
        std::size_t buf_minfill)
{
    while (!stop_flag.load())
    {
        if (buf->queued_samples() == 0)
        {
            // The buffer is empty. Perhaps the output stream is consuming
            // samples faster than we can produce them. Wait until the buffer
            // is back at its nominal level to make sure this does not happen
            // too often.
            buf->wait_buffer_fill(buf_minfill);
        }

        if (buf->pull_end_reached())
        {
            // Reached end of stream.
            break;
        }

        // Get samples from buffer and write to output.
        IQSampleVector samples = buf->pull();
        output->write(samples);

        if (!(*output))
        {
            fprintf(stderr, "ERROR: Output: %s\n", output->error().c_str());
        }
    }
}


/** Handle Ctrl-C and SIGTERM. */
static void handle_sigterm(int sig)
{
    stop_flag.store(true);

    std::string msg = "\nGot signal ";
    msg += strsignal(sig);
    msg += ", stopping ...\n";

    const char *s = msg.c_str();
    ssize_t r = write(STDERR_FILENO, s, strlen(s));

    if (r != (ssize_t) strlen(s)) {
        msg += " write incomplete";
    }
}


void usage()
{
    fprintf(stderr,
    "Usage: sdrdaemonrx [options]\n"
            "  -t devtype     Device type:\n"
#ifdef HAS_RTLSDR
            "                   - rtlsdr:  RTL-SDR devices\n"
#endif
#ifdef HAS_HACKRF
            "                   - hackrf:  HackRF One or Jawbreaker\n"
#endif
#ifdef HAS_AIRSPY
            "                   - airspy:  Airspy\n"
#endif
#ifdef HAS_BLADERF
            "                   - bladerf: BladeRF\n"
#endif
            "                   - test:    Test signal generator (CW carrier)\n"
            "  -c config      Startup configuration. Comma separated key=value configuration pairs\n"
            "                 or just key for switches. See below for valid values\n"
            "  -d devidx      Device index, 'list' to show device list (default 0)\n"
            "  -b blocks      Set buffer size in number of UDP blocks (default: 480 512 samples blocks)\n"
            "  -I address     IP address. Samples are sent to this address (default: 127.0.0.1)\n"
            "  -D port        Data port. Samples are sent on this UDP port (default 9090)\n"
            "  -C port        Configuration port (default 9091). The configuration string as described below\n"
            "                 is sent on this port via nanomsg in TCP to control the device\n"
            "\n"
            "Configuration options for the UDP sender:\n"
            "  txwait=<int>   Wait this number of microseconds (usleep) between transmission of each UDP packet (default 200)\n"
            "\n"
            "Configuration options for the decimator:\n"
            "  decim=<int>    log2 of decimation factor (default 0: no decimation)\n"
            "  fcpos=<int>    Center frequency position (default 2: center):\n"
            "                   - 0: Infradyne\n"
            "                   - 1: Supradyne\n"
            "                   - 2: Centered\n"
            "\n"
            "Configuration options for the Forward Erasure Correction:\n"
            "  fecblk=<int>   Number of additional FEC blocks (1..128, default 32)\n"
            "\n"
#ifdef HAS_RTLSDR
            "Configuration options for RTL-SDR devices\n"
            "  freq=<int>     Center frequency of operation in Hz (default 100000000)\n"
            "                 valid values: 10M to 2.2G (working range depends on device)\n"
            "  srate=<int>    Device sample rate in Hz (default 1000000)\n"
            "                 (valid ranges: [225001, 300000], [900001, 3200000]))\n"
            "  gain=<float>   Set LNA gain in dB, or 'auto',\n"
            "                 or 'list' to just get a list of valid values (default auto)\n"
            "  blklen=<int>   Set read buffer size in seconds (default RTL-SDR default)\n"
            "  ppmp=<int>     Set LO correction in positive PPM. Takes precedence over ppmn parameter (default 0)\n"
            "  ppmn=<int>     Set LO correction in negative PPM (default 0)\n"
            "  agc=<int>      1: Enable 0: disable RTL AGC mode (default 0: disabled)\n"
            "\n"
#endif
#ifdef HAS_HACKRF
            "Configuration options for HackRF devices\n"
            "  freq=<int>     Center frequency of operation in Hz (default 100000000)\n"
            "                 valid values: 1M to 6G\n"
            "  srate=<int>    Device sample rate in Hz (default 5000000)\n"
            "                 (valid ranges: [2500000,20000000]))\n"
            "  ppmp=<float>   Set LO correction in positive PPM. Takes precedence over ppmn parameter (default 0)\n"
            "  ppmn=<float>   Set LO correction in negative PPM (default 0)\n"
            "  lgain=<int>    LNA gain in dB. 'list' to just get a list of valid values: (default 16)\n"
            "  vgain=<int>    VGA gain in dB. 'list' to just get a list of valid values: (default 22)\n"
            "  bwfilter=<int> Filter bandwidth in MHz. 'list' to just get a list of valid values: (default 2.5)\n"
            "  extamp=<int>   1: Enable 0: disable extra RF amplifier (default 0: disabled)\n"
            "  antbias=<int>  1: Enable 0: disable antemma bias (default 0: disabled)\n"
            "\n"
#endif
#ifdef HAS_AIRSPY
            "Configuration options for Airspy devices\n"
            "  freq=<int>     Center frequency of operation in Hz (default 100000000)\n"
            "                 valid values: 24M to 1.8G\n"
            "  srate=<int>    Device sample rate in Hz. Depends on Airspy firmware and libairspy support\n"
            "                 Airspy firmware and library must support dynamic sample rate query. (default 10000000)\n"
            "  ppmp=<float>   Set LO correction in positive PPM. Takes precedence over ppmn parameter (default 0)\n"
            "  ppmn=<float>   Set LO correction in negative PPM (default 0)\n"
            "  lgain=<int>    LNA gain in dB. 'list' to just get a list of valid values: (default 8)\n"
            "  mgain=<int>    Mixer gain in dB. 'list' to just get a list of valid values: (default 8)\n"
            "  vgain=<int>    VGA gain in dB. 'list' to just get a list of valid values: (default 8)\n"
            "  antbias=<int>  1: Enable 0: disable antemma bias (default 0: disabled)\n"
            "  lagc=<int>     1: Enable 0: disable  LNA AGC (default 0: disabled)\n"
            "  magc=<int>     1: Enable 0: disable  mixer AGC (default 0: disabled)\n"
            "\n"
#endif
#ifdef HAS_BLADERF
            "Configuration options for BladeRF devices\n"
            "  freq=<int>     Center frequency of operation in Hz (default 300000000)\n"
            "                 valid values (with XB200): 100k to 3.8G\n"
            "                 valid values (without XB200): 300M to 3.8G\n"
            "  srate=<int>    Device sample rate in Hz. Valid values: 48k to 40M (default 1000000)\n"
            "  bw=<int>       Bandwidth in Hz. 'list' to just get a list of valid values: (default 1500000)\n"
            "  lgain=<int>    LNA gain in dB. 'list' to just get a list of valid values: (default 3)\n"
            "  v1gain=<int>   VGA1 gain in dB. 'list' to just get a list of valid values: (default 20)\n"
            "  v2gain=<int>   VGA2 gain in dB. 'list' to just get a list of valid values: (default 9)\n"
            "\n"
#endif
            "Configuration options for the test signal generator\n"
            "  freq=<int>     Center frequency sent in meta data in Hz. Valid values 10k to 10G (default 435000000)\n"
            "  srate=<int>    Sample rate sent in meta data in Hz. Valid values: 8k to 10M (default 5000000)\n"
            "  dfp=<int>      Positive shift frequency of carrier from center frequency in Hz (default 100000)\n"
            "  dfn=<int>      Negative shift frequency of carrier from center frequency in Hz (default 100000)\n"
            "  power=<int>    Signal peak power in negative dB. (default 0)\n"
            "\n");
}


void badarg(const char *label)
{
    usage();
    fprintf(stderr, "ERROR: Invalid argument for %s\n", label);
    exit(1);
}


bool parse_int(const char *s, int& v, bool allow_unit=false)
{
    char *endp;
    long t = strtol(s, &endp, 10);
    if (endp == s)
        return false;
    if ( allow_unit && *endp == 'k' &&
         t > INT_MIN / 1000 && t < INT_MAX / 1000 ) {
        t *= 1000;
        endp++;
    }
    if (*endp != '\0' || t < INT_MIN || t > INT_MAX)
        return false;
    v = t;
    return true;
}


static bool get_device(std::vector<std::string> &devnames, std::string& devtype, DeviceSource **srcsdr, int devidx)
{
    bool deviceDefined = false;
#ifdef HAS_RTLSDR
    if (strcasecmp(devtype.c_str(), "rtlsdr") == 0)
    {
        RtlSdrSource::get_device_names(devnames);
        deviceDefined = true;
    }
#endif
#ifdef HAS_HACKRF
    if (strcasecmp(devtype.c_str(), "hackrf") == 0)
    {
        HackRFSource::get_device_names(devnames);
        deviceDefined = true;
    }
#endif
#ifdef HAS_AIRSPY
    if (strcasecmp(devtype.c_str(), "airspy") == 0)
    {
        AirspySource::get_device_names(devnames);
        deviceDefined = true;
    }
#endif
#ifdef HAS_BLADERF
    if (strcasecmp(devtype.c_str(), "bladerf") == 0)
    {
        BladeRFSource::get_device_names(devnames);
        deviceDefined = true;
    }
#endif
    if (strcasecmp(devtype.c_str(), "test") == 0)
    {
        TestSource::get_device_names(devnames);
        deviceDefined = true;
    }

    if (!deviceDefined)
    {
        fprintf(stderr, "ERROR: wrong device type (-t option) must be one of the following:\n");
#ifdef HAS_RTLSDR
        fprintf(stderr, "       rtlsdr\n");
#endif
#ifdef HAS_HACKRF
        fprintf(stderr, "       hackrf\n");
#endif
#ifdef HAS_AIRSPY
        fprintf(stderr, "       airspy\n");
#endif
#ifdef HAS_BLADERF
        fprintf(stderr, "       bladerf\n");
#endif
        fprintf(stderr, "       test\n");
        return false;
    }

    if (devidx < 0 || (unsigned int)devidx >= devnames.size())
    {
        if (devidx != -1)
        {
            fprintf(stderr, "ERROR: invalid device index %d\n", devidx);
        }

        fprintf(stderr, "Found %u devices:\n", (unsigned int)devnames.size());

        for (unsigned int i = 0; i < devnames.size(); i++)
        {
            fprintf(stderr, "%2u: %s\n", i, devnames[i].c_str());
        }

        return false;
    }

    fprintf(stderr, "using device %d: %s\n", devidx, devnames[devidx].c_str());

#ifdef HAS_RTLSDR
    if (strcasecmp(devtype.c_str(), "rtlsdr") == 0)
    {
        // Open RTL-SDR device.
        *srcsdr = new RtlSdrSource(devidx);
    }
#endif
#ifdef HAS_HACKRF
    if (strcasecmp(devtype.c_str(), "hackrf") == 0)
    {
        // Open HackRF device.
        *srcsdr = new HackRFSource(devidx);
    }
#endif
#ifdef HAS_AIRSPY
    if (strcasecmp(devtype.c_str(), "airspy") == 0)
    {
        // Open Airspy device.
        *srcsdr = new AirspySource(devidx);
    }
#endif
#ifdef HAS_BLADERF
    if (strcasecmp(devtype.c_str(), "bladerf") == 0)
    {
        // Open BladeRF device.
        *srcsdr = new BladeRFSource(devnames[devidx].c_str());
    }
#endif
    if (strcasecmp(devtype.c_str(), "test") == 0)
    {
        // Open test device.
        *srcsdr = new TestSource(0);
    }

    return true;
}

int main(int argc, char **argv)
{
    int     devidx  = 0;
    std::string  filename;
    std::string  alsadev("default");
    std::string config_str;
    std::string devtype_str;
    std::vector<std::string> devnames;
    std::string dataaddress("127.0.0.1");
    unsigned int dataport = 9090;
    unsigned int cfgport = 9091;
    DeviceSource  *srcsdr = 0;
    unsigned int outputbuf_samples = 48 * UDPSIZE;
//    uint32_t compressedMinSize = 0;
//    bool useFec = true;
    unsigned int nbFECBlocks = 0;
    unsigned int txDelay = 0;

    fprintf(stderr,
            "SDRDaemonRx - Collect samples from SDR device and send it over the network via UDP\n");

    const struct option longopts[] = {
        { "devtype",    2, NULL, 't' },
        { "config",     2, NULL, 'c' },
        { "dev",        1, NULL, 'd' },
        { "buffer",     1, NULL, 'b' },
        { "daddress",   2, NULL, 'I' },
        { "dport",      1, NULL, 'D' },
        { "cport",      1, NULL, 'C' },
        { NULL,         0, NULL, 0 } };

    int c, longindex, value;
    while ((c = getopt_long(argc, argv,
            "t:c:d:b:I:D:C:",
            longopts, &longindex)) >= 0)
    {
        switch (c)
        {
            case 't':
                devtype_str.assign(optarg);
                break;
            case 'c':
                config_str.assign(optarg);
                break;
            case 'd':
                if (!parse_int(optarg, devidx))
                    devidx = -1;
                break;
            case 'b':
                if (!parse_int(optarg, value) || (value < 0)) {
                    badarg("-b");
                } else {
                    outputbuf_samples = value;
                }
                break;
            case 'I':
                dataaddress.assign(optarg);
                break;
            case 'D':
                if (!parse_int(optarg, value) || (value < 0)) {
                    badarg("-D");
                } else {
                    dataport = value;
                }
                break;
            case 'C':
                if (!parse_int(optarg, value) || (value < 0)) {
                    badarg("-C");
                } else {
                    cfgport = value;
                }
                break;
            default:
                usage();
                fprintf(stderr, "ERROR: Invalid command line options\n");
                exit(1);
        }
    }

    if (optind < argc)
    {
        usage();
        fprintf(stderr, "ERROR: Unexpected command line options\n");
        exit(1);
    }

    // Catch Ctrl-C and SIGTERM
    struct sigaction sigact;
    sigact.sa_handler = handle_sigterm;
    sigemptyset(&sigact.sa_mask);
    sigact.sa_flags = SA_RESETHAND;

    if (sigaction(SIGINT, &sigact, NULL) < 0)
    {
        fprintf(stderr, "WARNING: can not install SIGINT handler (%s)\n", strerror(errno));
    }

    if (sigaction(SIGTERM, &sigact, NULL) < 0)
    {
        fprintf(stderr, "WARNING: can not install SIGTERM handler (%s)\n", strerror(errno));
    }

    // Prepare output writer.
    UDPSink *udp_output_instance;
    udp_output_instance = new UDPSinkFEC(dataaddress, dataport);

//    if (useFec) {
//        udp_output_instance = new UDPSinkFEC(dataaddress, dataport);
//    } else if (compressedMinSize) {
//        udp_output_instance = new UDPSinkLZ4(dataaddress, dataport, UDPSIZE, compressedMinSize);
//    } else {
//        udp_output_instance = new UDPSinkUncompressed(dataaddress, dataport, UDPSIZE);
//    }

    std::unique_ptr<UDPSink> udp_output(udp_output_instance);

    if (!(*udp_output))
    {
        fprintf(stderr, "ERROR: UDP Output: %s\n", udp_output->error().c_str());
        exit(1);
    }

    if (!get_device(devnames, devtype_str, &srcsdr, devidx))
    {
        exit(1);
    }

    if (!(*srcsdr))
    {
        fprintf(stderr, "ERROR source: %s\n", srcsdr->error().c_str());
        delete srcsdr;
        exit(1);
    }

    //fprintf(stderr, (std::is_trivially_copyable<IQSample>::value ? "IQSample is trivially copiable\n" : "IQSample is NOT trivially copiable\n"));

    // Configure device and start streaming.

    srcsdr->setConfigurationPort(cfgport);

    // Prepare downsampler.
    Downsampler dn;
    srcsdr->associateDownsampler(&dn);

    if (!srcsdr->configure(config_str))
    {
        fprintf(stderr, "ERROR: source configuration: %s\n", srcsdr->error().c_str());
        delete srcsdr;
        exit(1);
    }

    /*
    if (!dn.configure(m))
    {
        fprintf(stderr, "ERROR: downsampler configuration: %s\n", dn.error().c_str());
        delete srcsdr;
        exit(1);
    }*/

    double freq = srcsdr->get_received_frequency();
    fprintf(stderr, "tuned for:         %.6f MHz\n", freq * 1.0e-6);

    double tuner_freq = srcsdr->get_frequency();
    fprintf(stderr, "device tuned for:  %.6f MHz\n", tuner_freq * 1.0e-6);

    double ifrate = srcsdr->get_sample_rate();
    fprintf(stderr, "IF sample rate:    %.0f Hz\n", ifrate);

    srcsdr->print_specific_parms();

    // Create source data queue.
    DataBuffer<IQSample> source_buffer;

    // ownership will be transferred to thread therefore the unique_ptr with move is convenient
    // if the pointer is to be shared with the main thread use shared_ptr (and no move) instead
    std::unique_ptr<DeviceSource> up_srcsdr(srcsdr);

    // Start reading from device in separate thread.
    //std::thread source_thread(read_source_data, std::move(up_srcsdr), &source_buffer);
    up_srcsdr->start(&source_buffer, &stop_flag);

    if (!up_srcsdr)
    {
        fprintf(stderr, "ERROR: source: %s\n", up_srcsdr->error().c_str());
        exit(1);
    }

    // If buffering enabled, start background output thread.
    DataBuffer<IQSample> output_buffer;
    std::thread output_thread;

    if (outputbuf_samples > 0)
    {
        output_thread = std::thread(write_output_data,
                               udp_output.get(),
                               &output_buffer,
                               outputbuf_samples);
    }

    IQSampleVector outsamples;
    bool inbuf_length_warning = false;

    // Main loop.
    for (unsigned int block = 0; !stop_flag.load(); block++)
    {

        // Check for overflow of source buffer.
        if (!inbuf_length_warning && source_buffer.queued_samples() > 10 * ifrate)
        {
            fprintf(stderr, "\nWARNING: Input buffer is growing (system too slow)\n");
            inbuf_length_warning = true;
        }

        // Pull next block from source buffer.
        IQSampleVector iqsamples = source_buffer.pull();

        if (iqsamples.empty())
        {
            break;
        }

        udp_output->setCenterFrequency(srcsdr->get_received_frequency());

        unsigned int confNbFECBlocks = srcsdr->get_nb_fec_blocks();

        if (confNbFECBlocks != nbFECBlocks)
        {
            nbFECBlocks = confNbFECBlocks;
            udp_output->setNbBlocksFEC(nbFECBlocks);
        }

        unsigned int confTxDelay = srcsdr->get_tx_delay();

        if (confTxDelay != txDelay)
        {
            txDelay = confTxDelay;
            udp_output->setTxDelay(txDelay);
        }

        // Possible downsampling and write to UDP

        if (dn.getLog2Decimation() == 0)
        {
            unsigned int sampleSize = srcsdr->get_sample_bits();
        	dn.rescale(sampleSize, iqsamples);

            udp_output->setSampleBits(srcsdr->get_sample_bits());
            udp_output->setSampleBytes((srcsdr->get_sample_bits()-1)/8 + 1);
            udp_output->setSampleRate(srcsdr->get_sample_rate());

            if (outputbuf_samples > 0)
            {
                // Buffered write.
                output_buffer.push(move(iqsamples));
            }
            else
            {
                // Direct write.
                udp_output->write(iqsamples);
            }
        }
        else
        {
            unsigned int sampleSize = srcsdr->get_sample_bits();
            dn.process(sampleSize, iqsamples, outsamples);

            udp_output->setSampleBits(sampleSize);
            udp_output->setSampleBytes((sampleSize -1)/8 + 1);
            udp_output->setSampleRate(srcsdr->get_sample_rate() / (1<<dn.getLog2Decimation()));

            // Throw away first block. It is noisy because IF filters
            // are still starting up.
            if (block > 0)
            {
                // Write samples to output.
                if (outputbuf_samples > 0)
                {
                    // Buffered write.
                    output_buffer.push(move(outsamples));
                }
                else
                {
                    // Direct write.
                    udp_output->write(outsamples);
                }
            }
        }
    }

    fprintf(stderr, "\n");

    // Join background threads.
    //source_thread.join();
    up_srcsdr->stop();

    if (outputbuf_samples > 0)
    {
        output_buffer.push_end();
        output_thread.join();
    }

    // No cleanup needed; everything handled by destructors

    return 0;
}

/* end */
